### PR TITLE
Hhomework week1

### PR DESCRIPTION
**Pull request by NelsonNaja**
Hello. THis is my homework. 

I tried to divide each process into files. Also I tried to implement some type of session :-) because I see a problem when you have two or more connections on the same port because they overlap, I tried to divide those connections.
 I read that manage conecctions it is possible to manage  with express but only with http I didt'n find. Maybe It is other topic differt to the homework but I seem me interesting. 

Thanks.

### CODE REVIEW

File: week1/homework/src/server.js [hunk 1]
```diff
@@ -1,20 +1,10 @@
...
- * will fail.
- */
-function createServer(port) {
-  let state = 10;
-
-  const server = http.createServer((request, response) => {
-    // TODO: Write your homework code here
-  });
-
+const handleRequest = require('./handleRequest');
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/server.js]: @NelsonNaja Nice. Good job on starting off by abstracting.

File: week1/homework/src/handleRequest.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+'use strict';
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/handleRequest.js]: @NelsonNaja Style Comment

Add a newline after the `use strict`.

File: week1/homework/src/handleRequest.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+const sendIndexPage = require('./responses/sendIndexPage');
+const sendError = require('./responses/sendError');
+const sendStyles = require('./responses/sendStyles');
+let session = {};
+
+function handleRequest(request, response) {
+  // I tried to avoid the error when you have two o more conexions in the
+  // same port, I created an id to identificate each conexion keeping
+  // the value status in a session object.
+  let id = request.headers.host + request.headers['user-agent'];
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/handleRequest.js]: @NelsonNaja Style Comment.

More newlines. The statements are a _little_ hard to read since they are very close together. Definitely put a newline before and after block, and open a block for if statements, even if they are a single line. Also, avoid ternaries unless they are very short. Three-level deep ternaries are a definite NO. With some effort, the logic of it can be extracted, but code is read ten times more often than written, so always think of: what would make this quick and easy to read 6 months later when all the context is lost?

File: week1/homework/src/handleRequest.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  let callback = urlEvaluation(request).callback;
+  if (session[id] === undefined) session[id] = DEFAULT_STATE;
+  urlEvaluation(request).urlEnable
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/handleRequest.js]: @NelsonNaja This is truly hard to parse at a glance. Although not strictly required to complete the homework, I'd recommend refactoring this. It looks like you're performing routing. I'd recommend just plain if-else blocks. In general avoid any kind of code that nests beyond a single level. And avoid ternaries as much as possible unless they are very short and simple (one line, few variables).

File: week1/homework/src/responses/sendError.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+function sendError(response) {
+  response.setHeader('Content-Type', 'application/json');
+  response.statusCode = 404;
+  response.write(`{
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/responses/sendError.js]: @NelsonNaja Prefer the practice of constructing a JS object first and then stringifying it via `JSON.stringify` instead of writing JSON objects by hand.

This has the added benefit that if you make a mistake, both your IDE and the interpreter will catch it easily.

So try doing:
```js
const error = {
  error: "Not found" 
};
response.write(JSON.stringify(error));
```

File: week1/homework/src/responses/sendIndexPage.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+function sendIndexPage(response) {
+  response.setHeader('Content-Type', 'text/html');
+  response.write(`
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/responses/sendIndexPage.js]: @NelsonNaja Good extra mile! 🙌

File: week1/homework/src/responses/setAdd.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+'use strict';
+const sendJson = require('./sendJson');
+function setAdd(response, session, id) {
+  session[id] += 1;
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/responses/setAdd.js]: @NelsonNaja One comment here would be consistency. I see in the subtraction you're doing `--` instead of `-= 1`. I'm not saying either method is better than the other, in this context, but stick to *one* of them.

File: week1/homework/src/responses/setReset.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+'use strict';
+const DEFAULT_STATE = 10;
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/responses/setReset.js]: @NelsonNaja It's usually a good idea to have constants like this in one place. We know that `DEFAULT_STATE` should have **one** value. But it's now been declared in two places. What if the requirements change you forget to change in both places?

Consider *exporting* the `DEFAULT_STATE` from one file and *using* it in multiple files.

File: week1/homework/src/urlEvaluation.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable max-len */
+'use strict';
+function urlEvaluation(request) {
+  const resUrl = request.url;
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/urlEvaluation.js]: @NelsonNaja a point of confusion here is that you're reading the`request.url` into a variable, but it's called `resUrl`, which sounds like `responseUrl`.

File: week1/homework/src/urlEvaluation.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+  return {
+    callback: resUrl.split('').filter(x => x !== '/').join(''),
+    urlRoot: resUrl === '/',
+    urlStyles: resUrl === '/styles.css',
+    urlEnable: resUrl === '/' || resUrl === '/state' || resUrl === '/add' || resUrl === '/subtract' || resUrl === '/reset' || resUrl === '/styles.css'
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/urlEvaluation.js]: @NelsonNaja Another, cleaner way to do this is to have an array of the supported urls:

```js
const supportedUrls = [
  '/',
  '/state',
  '/add',
  '/subtract',
  '/reset',
  '/style.css'
];
```
And then testing to see if the requested url is from that array.

```js
urlEnable: supportedUrls.includes(resUrl)
```

File: week1/homework/src/responses/sendStyles.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+function sendStyles(response) {
+  response.setHeader('Content-Type', 'text/css');
+  response.write(`
+    body {
+      background: #333333;
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/responses/sendStyles.js]: @NelsonNaja CSS is a weak point of mine, but I think you can shorten this to `#333` and `#DDD`. Be sure to confirm this, though.

File: week1/homework/src/urlEvaluation.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable max-len */
+'use strict';
+function urlEvaluation(request) {
+  const resUrl = request.url;
+  return {
+    callback: resUrl.split('').filter(x => x !== '/').join(''),
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/urlEvaluation.js]: @NelsonNaja A `callback` itself is usually _always_ a function and reading the name makes everyone think of a function first. But it seems like you're using this as in index to look up a function to execute. Consider renaming it.

Comment (BadgerBadgerBadgerBadger) [reply week1/homework/src/urlEvaluation.js]: @NelsonNaja Also comment on what exactly is going on here or why it is necessary. It might seem obvious to you but isn't to others, and won't be to you six months down the line.

File: week1/homework/src/getProcess.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+const setAdd = require('./responses/setAdd');
+const setSubtract = require('./responses/setSubtract');
+const setReset = require('./responses/setReset');
+const getState = require('./responses/getState');
+
+function getProcess(response, session, id) {
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/getProcess.js]: @NelsonNaja I would name it something other than `process`. A `process` in software engineering is already a specific concept (your NodeJS server is a process on the machine it is running), hence when you said you've created multiple _processes_ I was confused for a moment since that's some advanced stuff. You could simply call it `getStore` or something similar.

**Comment by BadgerBadgerBadgerBadger** [comment]
@NelsonNaja Left a bunch of comments. The homework itself is definitely good, but I'd like you to look at the feedback and consider refactoring for readability. Good job, overall! 🙌

File: week1/homework/src/handleRequest.js [hunk 14]
```diff
@@ -1,24 +1,39 @@
...
+    session[id] = DEFAULT_STATE;
+  }
+
+  if (urlEvaluation(request).urlEnable) {
+    if (urlEvaluation(request).urlRoot) {
+      sendIndexPage(response)
+    } else if (urlEvaluation(request).urlStyles) {
+      sendStyles(response)
+    } else {
+      getStore(response, session, id)[operation]()
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/handleRequest.js]: @NelsonNaja Nice. Definitely more appropriate name.

File: week1/homework/src/responses/sendError.js [hunk 15]
```diff
@@ -1,11 +1,13 @@
 'use strict';
 
+const error = {
+  error: "Not found"
+};
+
 function sendError(response) {
   response.setHeader('Content-Type', 'application/json');
+  response.write(JSON.stringify(error));
```
Comment (BadgerBadgerBadgerBadger) [inline week1/homework/src/responses/sendError.js]: @NelsonNaja Have a look at the tests. The 404 is not being returned correctly. Think about why.

Comment (NelsonNaja) [reply week1/homework/src/responses/sendError.js]: mmm. By the order maybe: it must be set  the statusCode first..

Comment (NelsonNaja) [reply week1/homework/src/responses/sendError.js]: I'm sorry for asking but I was thinking. Its important manage the sockets? or dont matter? or how must be do it ? ... for example in this homework  I saw a problem (see the pictures). I tried to use a type of idenficator (I named session) but obviosly its not apropiate.



I was reading https://nodejs.org/api/net.html#net_server_connections but  I'm stuck with that.

Maybe I'm confuse with that it is not so important.

Comment (BadgerBadgerBadgerBadger) [reply week1/homework/src/responses/sendError.js]: Heya. Don't be sorry for asking. It's what we're here for. A few points:

- I looked at the link you sent. I'd recommend to immediately ignore anything in red. 
image
We are currently in Node 12.0.0 and this was deprecated in 0.9.7, _many_ versions back.
- Your idea behind having different behavior for different users is the correct idea, except that the homework example we've provided doesn't have any user management system.Your diagram of the behavior is correct, but we already know and expect that this is a single user system and not a multi user system.

The thing is, when you go to a website like facebook, or almost any website, you are asked to sign in and you get an account. This sort of user management is not part of the homework and it's too early for it. Right now we don't have a mechanism for identifying a user. Also using sockets directly (which is very low level, by the way, you'll never practically need to do it).

Think of it instead like upvoting a post where all users  share the voting counter. Everyone sees the same and everyone's actions add to it.

Does that make more sense?

Comment (NelsonNaja) [reply week1/homework/src/responses/sendError.js]: Yes, that make sense. I already understand.. thanks. 🙂.

Comment (NelsonNaja) [reply week1/homework/src/responses/sendError.js]: Yes, that make sense. I already understand.. thanks. 🙂.

File: week1/homework/src/getStore.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+const setAdd = require('./responses/setAdd');
```
Comment (Buccaneer) [inline week1/homework/src/getStore.js]: Functions starting with `get` or `set` by convention only get the value of some object or set the value (which can contain some error checking). 
Using the name `setAdd` is a confusing here because there is no `add` object being set. Looking at the code for `setAdd` it actually modifies an object named `session` and then sends a JSON response. 
Try to describe what a function does instead, for example `increaseSessionStateAndRespond`. Don't be scared of longer names as long as they make your code more understandable.

Comment (Buccaneer) [reply week1/homework/src/getStore.js]: Or `addSessionStateAndRespond`, using increase is inconsistent on my part. That just shows how easy it is to introduce inconsistent naming ;)

File: week1/homework/src/handleRequest.js (hunk 14 above)
Comment (Buccaneer) [reply week1/homework/src/handleRequest.js]: `getStore` wraps the functions for add, subtract, reset and state behavior in one object which is useful but right now it is called every time you handle a request. So for every request it wraps those functions in a new object and returns that new object. Instead the wrapper object could just be created once beforehand and kept in a variable for later use.

I know that in this specific example it doesn't matter performance wise. Wrapping a few functions in an object is no effort but it's more about keeping in mind that some things can be done once and reused later.

File: week1/homework/src/getStore.js (hunk 16 above)
Comment (Buccaneer) [reply week1/homework/src/getStore.js]: Good work splitting up things in different modules. Keep in mind that there are extremes: reading and understanding code from one big file can be difficult, but many small files can be equally hard to piece together and understand. As a rule of thumb keep code that does one task together in one file until the task becomes too big then split it into smaller tasks.

In this instance I would merge the `setAdd`, `setSubtract`, `setReset` and `getState` modules into one module because these are really one task: modifying state. You could name it `state` because it contains all the behavior involving state. I noticed so far you've always exported single functions: `module.exports = someFunction` but you can export them wrapped in an object as well: `module.exports = { f1: function1, f2: function2 }`. If you export the `state` module like that you can have self explanatory code:
```js
const state = require('./state');
state.addAndRespond(); //notice how the function name can be shorter now because the module itself is called state
```

Comment (NelsonNaja) [reply week1/homework/src/getStore.js]: I got it thank you.
